kimchi/src/circuits/polynomials/foreign_field_add/witness.rs

```diff
@@ -267,90 +235,49 @@ fn init_foreign_field_fin_rows<F: PrimeField>(
-            WitnessCell::Standard(ZeroWitnessCell::create()),
-            WitnessCell::Standard(ZeroWitnessCell::create()),
-            WitnessCell::Standard(ZeroWitnessCell::create()),
-            WitnessCell::Standard(ZeroWitnessCell::create()),
-            WitnessCell::Standard(ZeroWitnessCell::create()),
-            WitnessCell::Standard(ZeroWitnessCell::create()),
-            WitnessCell::Standard(ZeroWitnessCell::create()),
-            WitnessCell::Standard(ZeroWitnessCell::create()),
-            WitnessCell::Standard(ZeroWitnessCell::create()),
-            WitnessCell::Standard(ZeroWitnessCell::create()),
+            CopyCell::create(bound_row, 0),     // bound_lo
+            CopyCell::create(bound_row + 1, 0), // bound_mi
+            CopyCell::create(bound_row + 2, 0), // bound_hi
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
         ],
     ];
 
-    for (row, wit) in witness_shape.iter().enumerate() {
-        for (col, cell) in wit.iter().enumerate() {
-            handle_ffadd_rows(
-                witness,
-                cell,
-                (row, col),
-                offset,
-                F::zero(),
-                F::zero(),
-                carry,
-            );
-        }
-    }
-}
-
-fn handle_ffadd_rows<F: PrimeField>(
-    witness: &mut [Vec<F>; COLUMNS],
-    witness_cell: &WitnessCell<F>,
-    coordinates: (usize, usize), /* row, col */
-    offset: usize,
-    sign: F,
-    overflow: F,
-    carry: [F; 2],
-) {
-    let (row, col) = coordinates;
-    match witness_cell {
-        WitnessCell::Standard(standard_cell) => {
-            handle_standard_witness_cell(
-                witness,
-                standard_cell,
-                offset + row,
-                col,
-                F::zero(), /* unused by this gate */
-            )
-        }
-        WitnessCell::FieldElement(elem_cell) => {
-            witness[col][offset + row] = {
-                match elem_cell.kind {
-                    FieldElementType::Overflow => overflow,
-                    FieldElementType::Carry => carry[elem_cell.limb_idx],
-                    FieldElementType::Sign => sign,
-                }
-            }
-        }
-        WitnessCell::Constant(field_elem) => witness[col][offset + row] = *field_elem,
-        WitnessCell::Ignore => (),
-    }
+    witness::init(
+        witness,
+        offset,
+        &witness_shape,
+        &variable_map![ "carry_lo" => carry[0], "carry_mi" => carry[1]],
```

Comment:
Interesting that you do not need to explain the rest of variables... It was annoying to set the other terms to zero in this section.

Reply:
Well, you only need to map whatever variables are used by the layout template.  In this case only two are used, so those are the only variables in the witness to set.

Reply:
Aha yeah that is very nice